test/ARCHITECTURE.md

```diff
@@ -0,0 +1,267 @@
+# Architecture
+
+This document describes the architecture of Cockpit's browser integration
+tests. The tests should replicate how a normal user interacts with Cockpit this
+requires a test machine which can easily add have multiple disks or interfaces,
+reboot, interact with multiple machines on the same network and run potentially
+destructive test scenario's (e.g. installing/updating packages, formatting
+disks).
+
+For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
+machine uses Cockpit specific virtual machine images maintained and created in
+the [bots](https://github.com/cockpit-project/bots) repository. The images are
+usually based on a distribution's cloud image customized with:
+
+* A well-known password for the admin/root user
+* Test SSH keys for access
+* Test packages required to test Cockpit
+* A build chroot with Cockpit's build dependencies to build the to be tested
+  Cockpit source inside the virtual machine offline. This allows a developer on
+  Fedora to easily prepare a Debian test image without having to install Debian
+  build tools.
+* Disabling system services which interfere with testing
+
+To replicate a user, Cockpit is tested in a browser controlled using the
+[Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
+(CDP) which is supported by Firefox and Chromium based browsers.
+
+The Python test framework in `test/common` is responsible for setting up the
+test environment, running tests and reporting of the test output.
+
+Diagram of the interaction of Browser/Machine/CDP/Test Framework.
+```mermaid
+graph TD;
+    id[Test Framework] <-->|CDP| Browser;
+    id[Test Framework] <-->|SSH| id1[Virtual Machine];
```

Comment:
perhaps a "cockpit" between browser and VM?

Reply:
That is a good suggestion. Not sure how we define the interaction between Cockpit and the VM though.

Reply:
I don't think what I wrote is ideal, need some more time to think of something better.